core/WASM_core.md

```diff
@@ -0,0 +1,352 @@
+WASI is quite immature, but *in theory* it can support these use cases.
+
+## WASM Runtime
+
+Regardless of the strategy, we need to choose a WASM runtime which each Lang SDK will use, so I'll
+address that first. The two major players in the space are
+[Wasmer](https://github.com/wasmerio/wasmer) and
+[Wasmtime](https://github.com/bytecodealliance/wasmtime). It's also not impossible that we might
+use one library from Rust for defining things like the host imported functions, but use different
+libraries for the runtime in the host langs.
+
+### Wasmer
+Wasmer's biggest pro is it's language support. Every language we currently have an SDK for (and
+more beyond that) has a Wasmer SDK for embedding and running WASM in that language. It's docs are
+decent, but not great.
+
+It has an MIT license.
+
+It's controlled by a startup (but hey, so are we) and has somewhat less active open-source
+development than Wasmtime.
+
+### Wasmtime
+Wasmtime supports all our languages _except_ Java, which is unfortunate, though it looks like
+there's an unofficial repo. It's also not immediately clear how usable JS/Node support is, 
+particularly when it comes to using WASI. It does have quite excellent docs, and built-in support
+for calling async functions from WASM, which is pretty awesome.
+
+It has an Apache 2.0 license.
+
+It's owned and maintained by the bytecode alliance, which also owns the WASM/WASI spec. It's backed
+by Mozilla.
+
+### Oh no CGo (and JNI)
+There's a big downside to both Wasmer and Wasmtime. Both of their Go implementations use CGo, and
+Wasmer's Java implementation uses JNI. This blows one of the bigger pluses out of the water.
+Luckily, [wazero](https://github.com/tetratelabs/wazero) exists, which we can use to get a pure Go
+environment. It supports WASI as well, but seemingly [not what we would need for network
+calls](https://wazero.io/specs/).
+
+I can't seem to find anything reasonable for Java. Maybe if we're blessed, such a thing
+will exist by the time we get around to porting Java, but that's a huge if.
+
+## Implementation Strategy
+
+### How much, if any WASI to use?
+
+The more we can lean on WASI, the less we have to expand the Core<-->Lang interface. That's a good
+thing in terms of both immediate work and long-term maintenance. There are some problems though:
+
+* Assuming we need to use `wazero` for non-CGo support, it doesn't appear to be able to handle
+  network calls. We'll test this, though. Even if we intend to use native gRPC clients, being able
+  to make network calls in Core is still pretty desirable for things like auto-updates, test server
+  downloads, etc.
+* Calling back into lang gives us opportunities for more user customization, which can be nice.
+  Especially when it comes to the metrics and tracing stuff. gRPC interceptors are also a potential
+  plus.
+
+My inclination for now is to at least *investigate* how much we can get out of WASI. Particularly
+with respect to running from Go.
```

Comment:
I don't think there's enough win for WASI. WASI has benefit if you need a POSIX-like abstraction which I don't think we do. Can just send extern/repr(C) structs in each direction as I'm doing with .NET. Many of the things we're doing are byte arrays or UTF-8 strings anyways.

Reply:
We *can*, of course, and probably will because WASI isn't very mature, but I think it's pretty much impossible to disagree with the statement that if I could compile core to WASM more or less as it is today and all the networking stuff worked, then we'd likely just do that. It saves on nearly all the work that needs to be done.

Now, that isn't the case, and it certainly seems like we'll have to go with passing back and forth, but if that assumption is wrong it's something worth knowing and worth spending a day to find out.

Reply:
:+1: There is no way for gRPC networking to work and for your goal of "Workers in the browser" to be accomplished, so the intermediary is required anyways. I am also hoping for a much smaller WASM blob when tonic, hyper, etc are not included.